lib/js/protocol/MessageFrameAssembler.js

```diff
@@ -30,55 +30,154 @@
+
 /**
- * This class will assemble the different types of messages
+ * Callback for finished packet.
+ * @callback cb
+ * @param {string} error - If there is an error that cannot be recovered from it will be returned.
+ * @param {SdlPacket} sdlPacket - Finished sdl packet.
  */
-class MessageFrameAssembler{
+
+/**
+ * Assembles sdl packets which may come in seperate chunks if the data exceeds the
+ * max transport unit allowed by sdl core.
+ * 
+ * @private {cb} _callback - Callback for combined packet.
+ * @private {number} _totalConsecutiveFrames - Total consecutive frames expected for a multiframe packet.
+ * @private {number} _consecutiveFramesDataLength - Expected data length of finished multiframe packet.
+ * @private {number} _consecutiveFramesHandledCount - Running total of consecutive frames received and handled.
+ * @private {array} _accumulator Combines incoming bytes from consecutive frames.
+ */
+class MessageFrameAssembler {
 
     /**
     * @constructor
     * @param {Number} headerSize
     * @param {Function} callback
     */
-    constructor(headerSize=8, callback){
-        if(callback == null){
+    constructor(callback) {
+        if (callback == null) {
             throw "Callback can't be null for MessageFrameAssembler";
         }
-        
-        this._accumulator = null; //TODO byte array output stream
-        this._headersize = headerSize;
-        this._totalSize = 0;
+
         this._callback = callback;
-    }
-    
-    /**
-    * @param {SdlPacket} sdlPacket
-    */
-    handleFrame(sdlPacket){
 
+        this._accumulator = [];
+        this._totalConsecutiveFrames = 0;
+        this._consecutiveFramesHandledCount = 0;
+        this._consecutiveFramesDataLength = 0;
     }
 
     /**
-    * @param {SdlPacket} sdlPacket
+     * Accepts a single frame or multiframe packet. The callback is used when the packet
+     * is completely read. For a single frame this is immediately. For a multi frame packet
+     * @param {SdlPacket} sdlPacket - Incoming sdl packet to be read.
     */
-    handleMultiFrameMessageFrame(sdlPacket){
-
+    handleFrame(sdlPacket) {
+        const frameType = sdlPacket.getFrameType();
+        //If the sdl packet is a single frame nothing needs to be assembled and this can be returned immediately.
+        if (frameType != FrameType.FIRST && frameType != FrameType.CONSECUTIVE) {
+            return this._callback(null, sdlPacket);
+        }
+        return this._handleMultiFrameMessage(sdlPacket);
     }
 
     /**
-    * @param {SdlPacket} sdlPacket
-    */
-    handleFirstDataFrame(sdlPacket){
+     * Handles the first data frame. Sets the expected frame count and data length.
+     * @param {SdlPacket} sdlPacket 
+     * @private
+     */
+    _handleFirstDataFrame(sdlPacket) {
+
+        /** @type {Uint8Array} */
+        let payload = sdlPacket.getPayload();
+
+        //first 4 8-bit integers contain the data length.
+        let dataLength = (payload[0] & 0x0F) << 24; //8-bit x 3
```

Comment:
Need a check that `payload` is not null and actually has data in it, including to the length we directly access the data in the array.